pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,110 @@
+ */
+
+package net.sourceforge.pmd.lang.apex.ast
+
+import com.google.summit.ast.*
+import com.google.summit.ast.declaration.*
+import com.google.summit.ast.modifier.KeywordModifier
+import com.google.summit.ast.modifier.Modifier
+import net.sourceforge.pmd.annotation.InternalApi
+import net.sourceforge.pmd.lang.apex.ApexParserOptions
+import net.sourceforge.pmd.lang.ast.ParseException
+import net.sourceforge.pmd.lang.ast.SourceCodePositioner
+
+@Deprecated("internal")
+@InternalApi
+@Suppress("DEPRECATION")
+class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    private val sourceCodePositioner = SourceCodePositioner(sourceCode)
+
+    /** Builds and returns an [ApexNode] AST corresponding to the given [root] node. */
+    fun build(root: CompilationUnit): ApexRootNode<TypeDeclaration> =
+        visit(root, parent = null) as? ApexRootNode<TypeDeclaration>
+            ?: throw ParseException("Unable to build tree")
+
+    /**
+     * Visits [node] and returns an [ApexNode] wrapper.
+     *
+     * Sets the parent of the resulting [ApexNode] to [parent], if it's not `null`.
+     */
+    private fun visit(node: Node?, parent: ApexNode<*>?): AbstractApexNode<*>? {
+        val wrapper: AbstractApexNode<*>? =
+            when (node) {
+                null -> null
+                is CompilationUnit -> visit(node.typeDeclaration, parent)
+                is TypeDeclaration -> visitTypeDeclaration(node)
+                is Identifier,
+                is KeywordModifier,
+                is TypeRef -> null
+                else -> {
+                    println("No adapter exists for type ${node::class.qualifiedName}")
+                    // TODO(b/239648780): temporary print
+                    null
+                }
+            }
+
+        wrapper?.setParent(parent)
+        wrapper?.handleSourceCode(sourceCode)
+        wrapper?.calculateLineNumbers(sourceCodePositioner)
+        return wrapper
+    }
+
+    /** Calls [visit] on each of [nodes]. */
+    private fun visit(nodes: List<Node>, parent: ApexNode<*>?) = nodes.forEach { visit(it, parent) }
+
+    /** Calls [visit] on each [child][Node.getChildren] of [node]. */
+    private fun visitAll(node: Node, parent: ApexNode<*>?) =
+        node.getChildren().forEach { visit(it, parent) }
+
+    /**
+     * Calls [visit] on each [child][Node.getChildren] of [node] that isn't already [visited].
+     *
+     * [visited] is a predicate that, given a child node `child`, should return `true` when `child` is
+     * equal to a property of [node] that was already visited and shouldn't be visited again.
+     */
+    private fun visitRest(node: Node, parent: ApexNode<*>?, visited: (Node) -> Boolean) =
+        node.getChildren().filterNot(visited).forEach { visit(it, parent) }
+
+    /** Visits the [TypeDeclaration] node and returns an [ApexRootNode] wrapper. */
+    private fun visitTypeDeclaration(node: TypeDeclaration) =
```

Comment:
The top-level `visit` method has a contract that it will do everything: recurse on children, setParent, etc.

Could we use a different name for the methods that do a subset of that (rather than more specific versions of `visit`)?  (Ideally not "visitHelper" 😄 )

Reply:
If I make the secondary `visit*` methods only construct a new wrapper (including building child nodes), could we keep the names then? In that case, `visit` is the only method calling `setParent` implicitly, and all other `visit*` methods perform a consistent & expected set of operations.

(If it helps, I am also thinking of replacing the name `visit` with `build`.)

Reply:
Sure. I'm just aiming for consistency in that whatever set of `[prefix]*` methods will take corresponding parameters and perform corresponding actions.